### PR TITLE
Validate block transactions

### LINKED ISSUE #519: Function that validates any kind of transaction

### LINKED ISSUE #520: Reorder validations as discussed in #488

### PR DESCRIPTION
**Pull request by lrubiorod**
Created a method to validate any kind of transaction
Created a method to validate block transactions and update utxo, dr_pool and tx_pool
Handled the number of commits or reveals in a block
Reordered validations

Implemented RadonMixed::ToArray
Implemented Array::get
Implemented RadonString::ToFloat

Close #519 
Close #520

### CODE REVIEW

File: validations/src/validations.rs [hunk 1]
```diff
@@ -386,38 +393,99 @@ pub fn validate_transaction_signatures(
...
 pub fn validate_transactions(
     utxo_set: &UnspentOutputsPool,
     _txn_pool: &TransactionsPool,
     data_request_pool: &DataRequestPool,
     block: &Block,
 ) -> Result<BlockInChain, failure::Error> {
-    // TODO: Add validate_mint function
+    let transactions = block.txns.clone();
+
+    match transaction_tag(&transactions[0].body) {
```
Comment (tmpolaczyk) [inline validations/src/validations.rs]: This will panic if transactions is an empty vector

**Comment by tmpolaczyk** [comment]
Data request for testing:

```json
{"jsonrpc":"2.0","method":"inventory","params":{"transaction":{"body":{"version":0,"inputs":[{"ValueTransfer":{"transaction_id":"d5789236a5da77528ec6e89c2dae22cb6b57abcc6ce9dfc6bada5d66bebad81b","output_index":0}}],"outputs":[{"DataRequest":{"pkh":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"data_request":{"not_before":0,"retrieve":[{"kind":"HTTP-GET","url":"https://openweathermap.org/data/2.5/weather?id=2950159&appid=b6907d289e10d714a6e88b30761fae22","script":[150, 83, 204, 132, 146, 1, 164, 109, 97, 105, 110, 204, 132, 146, 1, 164, 116, 101, 109, 112, 204, 130]},{"kind":"HTTP-GET","url":"https://api.apixu.com/v1/current.json?key=297bc8f9aa7841d7a0e95208180310&q=Berlin,DE","script":[150, 83, 204, 132, 146, 1, 167, 99, 117, 114, 114, 101, 110, 116, 204, 132, 146, 1, 166, 116, 101, 109, 112, 95, 99, 204, 130]}],"aggregate":{"script":[145,  146,  102,  32]},"consensus":{"script":[145,  146,  102,  32]},"deliver":[{"kind":"HTTP-GET","url":"https://hooks.zapier.com/hooks/catch/3860543/l2awcd/"},{"kind":"HTTP-GET","url":"https://hooks.zapier.com/hooks/catch/3860543/l1awcw/"}]},"value":1000,"witnesses":1,"backup_witnesses":1,"commit_fee":0,"reveal_fee":0,"tally_fee":0,"time_lock":0}}]},"signatures":[{"signature":{"Secp256k1":{"r":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"s":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"v":0}},"public_key":{"compressed":0,"bytes":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}}]}},"id":"1"} 
```

**Comment by tmpolaczyk** [comment]
Data request for testing: this one uses a random number API:

```json
{"jsonrpc":"2.0","method":"inventory","params":{"transaction":{"body":{"version":0,"inputs":[{"ValueTransfer":{"transaction_id":"d5789236a5da77528ec6e89c2dae22cb6b57abcc6ce9dfc6bada5d66bebad81b","output_index":0}}],"outputs":[{"DataRequest":{"pkh":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"data_request":{"not_before":0,"retrieve":[{"kind":"HTTP-GET","url":"http://qrng.anu.edu.au/API/jsonI.php?length=1&type=uint8","script":[149, 83, 204, 132, 146, 1, 164, 100, 97, 116, 97, 204, 128, 146, 1, 0]}],"aggregate":{"script":[145,  146,  102,  32]},"consensus":{"script":[145,  146,  102,  32]},"deliver":[{"kind":"HTTP-GET","url":"https://hooks.zapier.com/hooks/catch/3860543/l2awcd/"},{"kind":"HTTP-GET","url":"https://hooks.zapier.com/hooks/catch/3860543/l1awcw/"}]},"value":4000,"witnesses":4,"backup_witnesses":1,"commit_fee":0,"reveal_fee":0,"tally_fee":0,"time_lock":0}}]},"signatures":[{"signature":{"Secp256k1":{"r":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"s":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"v":0}},"public_key":{"compressed":0,"bytes":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}}]}},"id":"1"} 
```

File: validations/src/validations.rs [hunk 2]
```diff
@@ -1,17 +1,22 @@
...
+        Block, BlockError, BlockInChain, CheckpointBeacon, Epoch, Hash, Hashable, Input,
+        KeyedSignature, Output, OutputPointer, RADRequest, Signature, Transaction, TransactionBody,
+        TransactionError, TransactionType, TransactionsPool, UnspentOutputsPool,
     },
     data_request::DataRequestPool,
     serializers::decoders::{TryFrom, TryInto},
 };
 
-use witnet_crypto::signature::verify;
+use log::debug;
```
Comment (anler) [inline validations/src/validations.rs]: _Suggestion_: import just log `use log` and then use `log::debug(...)` etc. This way with a grep `log::` you can find all the places where we log.

File: validations/src/validations.rs [hunk 3]
```diff
@@ -24,14 +29,40 @@ pub fn transaction_inputs_sum(
...
-    for input in &tx.inputs {
-        let pointed_value = pool
-            .get(&input.output_pointer())
-            .ok_or_else(|| TransactionError::OutputNotFound {
-                output: input.output_pointer(),
-            })?
-            .value();
-        total_value += pointed_value;
+    match transaction_tag(tx) {
+        // Commit exception to calculate inputs
```
Comment (anler) [inline validations/src/validations.rs]: _Suggestion_: remove this kind of comment, they add almost 0 value

File: validations/src/validations.rs [hunk 4]
```diff
@@ -24,14 +29,40 @@ pub fn transaction_inputs_sum(
...
+            for input in &tx.inputs {
+                let pointed_value = pool
+                    .get(&input.output_pointer())
+                    .ok_or_else(|| TransactionError::OutputNotFound {
+                        output: input.output_pointer(),
+                    })?
+                    .value();
+                total_value += pointed_value;
+            }
+        }
```
Comment (anler) [inline validations/src/validations.rs]: _Suggestion_: the match expression is too big, an improvement could be to move into their own functions the logic of each match arm

File: validations/src/validations.rs [hunk 5]
```diff
@@ -129,14 +162,20 @@ pub fn validate_consensus(
     }
 }
 
+/// Function to validate a value transfer transaction
```
Comment (anler) [inline validations/src/validations.rs]: (***) _Suggestion_: this kind of comment add almost no insight on what the function does and transmits the same information as the name of the function. A better way would be to describe what does it mean for a value-transfer transaction to be valid and in which situations that might fail

File: validations/src/validations.rs [hunk 6]
```diff
@@ -129,14 +162,20 @@ pub fn validate_consensus(
...
+pub fn validate_vt_transaction(_tx: &TransactionBody) -> Result<(), failure::Error> {
+    // TODO(#514): Implement valu transfer validation
+    Ok(())
+}
+
 /// Function to validate a data request transaction
```
Comment (anler) [inline validations/src/validations.rs]: Same as above (***)

File: validations/src/validations.rs [hunk 7]
```diff
@@ -161,24 +200,21 @@ pub fn validate_dr_transaction(tx: &TransactionBody) -> Result<(), failure::Erro
 
 // Add 1 in the number assigned to a DataRequestOutput
 pub fn update_count<S: ::std::hash::BuildHasher>(
```
Comment (anler) [inline validations/src/validations.rs]: _Suggestion_: a better name would be `inc_witnesses` (inc for increment/increase) or something that better describes the actual behavior

File: validations/src/validations.rs [hunk 8]
```diff
@@ -161,24 +200,21 @@ pub fn validate_dr_transaction(tx: &TransactionBody) -> Result<(), failure::Erro
...
-        }
-        None => {
-            hm.insert(k.clone(), 1);
-        }
-    };
+    hm.entry(k.clone()).or_insert((0, rf)).0 += 1;
 }
 
+/// HashMap to count commit and reveals transactions need for a Data Request
+pub type WitnessesCounter<S> = HashMap<OutputPointer, (u32, u32), S>;
```
Comment (anler) [inline validations/src/validations.rs]: _Suggestion_: describe what represents each field of the tuple, or use a struct
```rust
pub struct WitnessesCount {
  commits: u32,
  reveals: u32
}
pub type WitnessesCounter<S> = HashMap<OutputPointer, WitnessesCount, S>;
```

Also, why do you parameterize with `S`?

File: validations/src/validations.rs [hunk 9]
```diff
@@ -386,38 +430,104 @@ pub fn validate_transaction_signatures(
...
 /// Function to validate transactions in a block and update a utxo_set and a `TransactionsPool`
-// TODO: Add verifications related to data requests (e.g. enough commitment transactions for a data request)
 pub fn validate_transactions(
     utxo_set: &UnspentOutputsPool,
     _txn_pool: &TransactionsPool,
     data_request_pool: &DataRequestPool,
     block: &Block,
 ) -> Result<BlockInChain, failure::Error> {
-    // TODO: Add validate_mint function
+    let transactions = block.txns.clone();
```
Comment (anler) [inline validations/src/validations.rs]: How expensive is this clone operation?

File: validations/src/validations.rs [hunk 10]
```diff
@@ -386,38 +430,104 @@ pub fn validate_transaction_signatures(
 }
 
 /// Function to validate a transaction
```
Comment (anler) [inline validations/src/validations.rs]: See (***)

File: validations/src/validations.rs [hunk 11]
```diff
@@ -386,38 +430,104 @@ pub fn validate_transaction_signatures(
...
+            debug!("DataRequest Transaction validation");
+            let fee = transaction_fee(transaction_body, utxo_set)?;
+
+            validate_tally_transaction(transaction_body, dr_pool, utxo_set, fee)?;
+            Ok(fee)
+        }
+    }
 }
 
 /// Function to validate transactions in a block and update a utxo_set and a `TransactionsPool`
```
Comment (anler) [inline validations/src/validations.rs]: See (***)

File: validations/src/validations.rs [hunk 12]
```diff
@@ -129,14 +162,20 @@ pub fn validate_consensus(
     }
 }
 
+/// Function to validate a value transfer transaction
+pub fn validate_vt_transaction(_tx: &TransactionBody) -> Result<(), failure::Error> {
+    // TODO(#514): Implement valu transfer validation
```
Comment (aesedepece) [inline validations/src/validations.rs]:
```suggestion
    // TODO(#514): Implement value transfer transaction validation
```

File: validations/src/validations.rs (hunk 7 above)
Comment (aesedepece) [reply validations/src/validations.rs]: I'm totally in for forceful expressive function names like `increment_witnesses_counter`.

File: validations/src/validations.rs (hunk 3 above)
Comment (lrubiorod) [reply validations/src/validations.rs]: :crying_cat_face:

File: validations/src/validations.rs (hunk 9 above)
Comment (aesedepece) [reply validations/src/validations.rs]: How _necesary_ is it?

File: validations/src/validations.rs (hunk 8 above)
Comment (lrubiorod) [reply validations/src/validations.rs]: Clippy likes the S

Comment (aesedepece) [reply validations/src/validations.rs]: For reference: https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#409-416

File: validations/src/validations.rs (hunk 9 above)
Comment (lrubiorod) [reply validations/src/validations.rs]: Modified by a reference

File: validations/src/validations.rs (hunk 5 above)
Comment (lrubiorod) [reply validations/src/validations.rs]: Created issue #550

**Pull request merged**